File: src/publisher.cpp
#include "ros/ros.h"
#include <performance_tests/SuperAwesome.h>
#include <sstream>
#include <string>

int main(int argc, char **argv)
{
  ros::init(argc, argv, "pub");

  std::string rate = "10";
  std::string nbTests = "50";
  
  if (argc > 3 || argc <= 0)
  {
    ROS_ERROR("FAIL");
    ROS_ERROR("usage: rosrun performance_tests pub [rate] [nbTests]");
    return 0;
  } 
  else if ( argc == 1) 
  {
    ROS_INFO("no rate nor nbTests defined, using default rate: [%s] hz and default nbTests: [%s]", rate.c_str(), nbTests.c_str());
  }
    else if ( argc == 2) 
  {
    ROS_INFO("no nbTests defined, using default nbTests: [%s]", nbTests.c_str(), nbTests.c_str());
    rate = argv[1];
  }
  else
  {
      rate = argv[1];
      nbTests = argv[2];
  }
  

  int iRate = atoi(rate.c_str());
  int iNbTests = atoi(nbTests.c_str());


  ros::NodeHandle n;

  ros::Publisher p = n.advertise<performance_tests::SuperAwesome>("SuperTopic", 1000);

  ros::Rate loop_rate(iRate);

  int count = 0;
  while (ros::ok() && count < iNbTests)
  {
    performance_tests::SuperAwesome msg;

    std::stringstream ss;
    ss << "nb " << count;
    msg.s = ss.str();

    //ROS_INFO("%s", msg.s.c_str());
    p.publish(msg);

    ros::spinOnce();

    loop_rate.sleep();
    ++count;
  }
  //sleep(1);
  
  performance_tests::SuperAwesome endMsg;
  endMsg.s = "FIN|Cpp|"+rate;

  //ROS_INFO("%s", endMsg.s.c_str());
  p.publish(endMsg);

  ros::spinOnce();
  loop_rate.sleep();


  return 0;
}

File: scripts/publisher.py
#!/usr/bin/env python
import rospy
from performance_tests.msg import SuperAwesome
import sys


def talker():
    freq = "10"
    nbTests = "50"
    if (len(sys.argv) > 3 or len(sys.argv) <= 0):
        rospy.loginfo("FAIL")
        rospy.loginfo("usage: rosrun performance_tests publisher.py [rate] [nbTests]")
    elif (len(sys.argv) == 1):
        rospy.loginfo("no rate nor nbTests defined, using default rate: [%s] hz and default nbTests: [%s]" % freq, nbTests)
    elif len(sys.argv) == 2:
        rospy.loginfo("no nbTests defined, using default nbTests: [%s]" % nbTests)
        freq = sys.argv[1]
    else:
        freq = sys.argv[1]
        nbTests = sys.argv[2]
    iRate = int(freq)
    iNbTests = int(nbTests)
    pub = rospy.Publisher('SuperTopic', SuperAwesome, queue_size=10)
    rospy.init_node('talker', anonymous=True)
    rate = rospy.Rate(iRate) 
    count = 0
    while not rospy.is_shutdown() and count < iNbTests:
        msg = "nb %s" % count #rospy.get_time()
        #rospy.loginfo(msg)
        pub.publish(msg)
        rate.sleep()
        count += 1
        
    EndMsg = "FIN|Python|"+freq
    #rospy.loginfo(EndMsg)
    pub.publish(EndMsg)
    rate.sleep()

if __name__ == '__main__':
    try:
        talker()
    except rospy.ROSInterruptException:
        pass
